### PR TITLE
xplorer: init at unstable-2022-08-11

### PR DESCRIPTION
**Pull request by dit7ya**
###### Description of changes

xplorer is a cross-platform file-manager written using the Tauri platform. My main motivation for packaging this was to figure out how difficult packaging Tauri apps would be. Turns out it is pretty simple - Tauri backends can be built using standard rust build-support in nixpkgs, and the frontend can be built using mkYarnPackage. 

I have not tested the bin outside x86_64 though.

https://github.com/kimlimjustin/xplorer

https://github.com/tauri-apps/tauri

**QUESTION**: Given that upstream provides both `package.json` and `yarn.lock`, can we not skip the `package.json` or the `offlineCache` ? It seems like it does not work without those.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dit7ya** [comment]
Result of `nixpkgs-review pr 187547` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>xplorer</li>
  </ul>
</details>

File: pkgs/applications/file-managers/xplorer/default.nix [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    substituteInPlace tauri.conf.json --replace '"distDir": "../out/src",' '"distDir": "frontend-build/src",'
+  '';
+
+  buildInputs = [ dbus openssl freetype libsoup gtk3 webkitgtk cmake ];
+  nativeBuildInputs = [ pkg-config ];
+
+  # Skip one test that fails ( tries to mutate the parent directory )
+  checkFlags = [ "--skip=test_file_operation" ];
+
+  # Rename the executable
```
Comment (SuperSandro2000) [inline pkgs/applications/file-managers/xplorer/default.nix:70]:
```suggestion
```

File: pkgs/applications/file-managers/xplorer/default.nix [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    cp -R ${frontend-build}/src frontend-build
+
+    substituteInPlace tauri.conf.json --replace '"distDir": "../out/src",' '"distDir": "frontend-build/src",'
+  '';
+
+  buildInputs = [ dbus openssl freetype libsoup gtk3 webkitgtk cmake ];
+  nativeBuildInputs = [ pkg-config ];
+
+  # Skip one test that fails ( tries to mutate the parent directory )
+  checkFlags = [ "--skip=test_file_operation" ];
```
Comment (SuperSandro2000) [inline pkgs/applications/file-managers/xplorer/default.nix:68]:
```suggestion
  checkFlags = [
    # tries to mutate the parent directory
    "--skip=test_file_operation"
  ];
```

File: pkgs/applications/file-managers/xplorer/default.nix [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+rustPlatform.buildRustPackage rec {
+  inherit version src pname;
+
+  sourceRoot = "source/src-tauri";
+
+  cargoSha256 = "sha256-U97V3b/luAsGxSFszwSNu0fvd6PpieDgF1BVECH2pMQ=";
+
+  # Copy the frontend static resources to final build directory
+  # Also modify tauri.conf.json so that it expects the resources at the new location
+  preBuild = ''
```
Comment (SuperSandro2000) [inline pkgs/applications/file-managers/xplorer/default.nix:57]:
```suggestion
  postPatch = ''
```

File: pkgs/applications/file-managers/xplorer/default.nix [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    mkdir -p frontend-build
+    cp -R ${frontend-build}/src frontend-build
```
Comment (SuperSandro2000) [inline pkgs/applications/file-managers/xplorer/default.nix:59]: Would symlinking work here?

**Comment by happysalada** [comment]
@dit7ya hey,, I think it would be really nice to have an example of an app packaged with tauri in nixpkgs. Are you still interested in getting this merged ?
I think it only needs a rebase and to address a few commands, I would be happy to help you get this merged.

**Comment by happysalada** [comment]
Ok there is another package that uses tauri that im interested in. Im going to give a go to that one and will ping you if i succeed.
Thanks a lot for laying the groundwork in this PR!

**Comment by dit7ya** [comment]
@happysalada  Actually, it worked after a while - I have updated the package! 

I would like to get this PR merged.

**Comment by happysalada** [comment]
Nice work !

**Pull request merged**